pkg/resources/management.cattle.io/v3/clusterproxyconfig/validator_test.go

```diff
@@ -0,0 +1,94 @@
+package clusterproxyconfig
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	v3api "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/webhook/pkg/admission"
+	wranglerfake "github.com/rancher/wrangler/pkg/generic/fake"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	admissionv1 "k8s.io/api/admission/v1"
+	authenicationv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var (
+	cpcGVR = metav1.GroupVersionResource{Group: "management.cattle.io", Version: "v3", Resource: "clusterproxyconfigs"}
+	cpcGVK = metav1.GroupVersionKind{Group: "management.cattle.io", Version: "v3", Kind: "ClusterProxyConfig"}
+)
+
+func Test_admitter_Admit(t *testing.T) {
+	tests := []struct {
+		name          string
+		alreadyExists bool
+		allowed       bool
+		wantErr       bool
+	}{
+		{
+			name:          "create clusterproxyconfig when none exists",
+			allowed:       true,
+			alreadyExists: false,
+		},
+		{
+			name:          "attempt to make more than one clusterproxyconfig",
+			allowed:       false,
+			alreadyExists: true,
+		},
+		{
+			name:    "failed to list clusterproxyconfigs",
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			cpsCache := wranglerfake.NewMockCacheInterface[*v3api.ClusterProxyConfig](ctrl)
+			cpsCache.EXPECT().List(gomock.Any(), gomock.Any()).DoAndReturn(func(namespace string, _ labels.Selector) ([]*v3api.ClusterProxyConfig, error) {
+				if tt.wantErr {
+					return nil, fmt.Errorf("simulated list error")
+				}
+				if tt.alreadyExists {
+					return []*v3api.ClusterProxyConfig{
+						{
+							Enabled: true,
+						},
+					}, nil
+				}
+				return nil, nil
+			}).AnyTimes()
+			a := &admitter{
+				cpsCache: cpsCache,
+			}
```

Comment:
I believe you don't need to re-create your cache, admitter, and other things for every subtest in the loop.

Reply:
Good points.  Thanks.

Reply:
FWIW, re-creating the cache, admitter, resolvers, etc. is how we've done it for many other resources. For example, GlobalRoles creates a new default state in each test

https://github.com/rancher/webhook/blob/6d6130f5fe1f70db905a851225d4929855e6dcf2/pkg/resources/management.cattle.io/v3/globalrole/validator_test.go#L726

https://github.com/rancher/webhook/blob/6d6130f5fe1f70db905a851225d4929855e6dcf2/pkg/resources/management.cattle.io/v3/globalrole/setup_test.go#L256